content/posts/ai-crawlers-abuse.md

```diff
@@ -0,0 +1,177 @@
+    x-amz-request-id: Z502YT87WEMM3ZY9
+    last-modified: Thu, 11 Feb 2021 09:12:59 GMT
+    etag: "c8cb418f5a8ff2e376fc5f7b7564e445"
+    x-amz-meta-mtime: 1495712654.422637991
+    accept-ranges: bytes
+    x-served: Nginx-Proxito-Sendfile
+    x-backend: web-i-01ce033e08bb601ef
+    referrer-policy: strict-origin-when-cross-origin
+    x-frame-options: DENY
+    x-content-type-options: nosniff
+    content-security-policy: object-src 'none'; frame-ancestors 'none'
+    cf-cache-status: DYNAMIC
+    set-cookie: __cf_bm=qGTC_35C03_QI6rw.JyPZhFHpo2QxUy7DMMFJpjz2_U-1721924658-1.0.1.1-4iS9rZHPJmt_I5rQX4NuKr_pHQmCw0jvCzAIYX.CeUtGZh6hIZIjBWhlPoxEMjhsRcvbuTSpgSa9oltlvYDtEA; path=/; expires=Thu, 25-Jul-24 16:54:18 GMT; domain=.readthedocs.org; HttpOnly; Secure; SameSite=None
+    set-cookie: _cfuvid=buQRaZWXJEn51CIpRsDW3E52DyDqEHd_sgY5PxOLfyE-1721924658779-0.0.1.1-604800000; path=/; domain=.readthedocs.org; HttpOnly; Secure; SameSite=None
+    server: cloudflare
+    cf-ray: 8a8d7fdbbf787565-SEA
+
+</code>
+
+As you can see, this file was last modified in 2021,
+but it was downloaded hundreds of times.
+
+</details>
+
+### 10 TB in June 2024 from another
+
+In June 2024, **someone used Facebook's content downloader to download 10 TB** of data, mostly htmlzip and PDF files. When we tried to email Facebook about it with the [contact information listed in the bot's user agent](http://www.facebook.com/externalhit_uatext.php), but the email bounced.
+
+![AI Crawler abuse, June 2024](/images/posts/bandwidth-june-2024.png)
+
+We think this was someone using Facebook's content downloader to download all of our files, but we aren't really sure.
+It would also be Meta crawling the site for their latest AI project,
+but the user agent is not clear.
+
+## Actions taken
+
+We have IP-based rate limiting in place for many of our endpoints,
+however these crawlers are coming from a large number of IP addresses,
+so our rate limiting is not effective.
+
+These crawlers are using real user agents,
+which is a good thing and allows us to identify them,
+but it's very difficult for us to block on user agent because many real users use the same browsers with the same user agent.
+
+We have taken a few actions to try to mitigate this abuse:
+
+* We have temporarily blocked all traffic from bots [Cloudflare identifies as AI Crawlers](https://radar.cloudflare.com/traffic/verified-bots), while we figure out how to deal with this.
+* We have reconfigured our CDN to better cache these files, so that our origin servers are not hit as hard.
+* We have started monitoring our bandwidth usage more closely and are working on more aggressive rate limiting rules.
+
+## Outcomes
+
+Given that our Community site is only for hosting open source projects,
+AWS and Cloudflare do give us sponsored plans,
+but we only have a limited number of credits each year.
+These bandwidth costs mean that we will run out of AWS credits.
```

Comment:
Because it's been a few paragraphs since we discussed bandwidth costs:

```suggestion
The additional bandwidth costs AI crawlers are currently causing will likely mean we will run out of AWS credits early.
```